### ISSUE TITLE
Add min/max buffer for LB listener rule priority

### ISSUE THREAD

**Issue by garyrutland** [hashicorp/terraform-provider-aws#22991]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

We currently use one load balancer with multiple domains/subdomains which are also split out into paths but have an issue where we want some rules to auto generate it's priority so that we don't need to manually keep track of what the next rule is but with others would need to set a specific priority.

An example of this would be if you were trying to split up a monolith into smaller services.

Based on the current docs this isn't possible.
https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lb_listener_rule#priority
> Leaving it unset will automatically set the rule with next available priority after currently existing highest rule.
This means if I set a priority to 50000, the next rule we try to add would attempt 50001 when in reality the next available could be 1 (assuming there is only the default and 50000 rule).

In the potential config below I've used `priority_min` and `priority_max` but these can be renamed if a bit confusing.
I've also set a `priority_min` although it isn't needed in this example, however, it shows how it could work when included and if you needed to inject something to be a high priority at a later date.
Also, for backwards compatibility, these can be set to 1 and 50000 respectively or if the feature is not needed.

### New or Affected Resource(s)

* aws_lb_listener_rule

### Potential Terraform Configuration

```hcl
locals {
  load_balancer_listener_arn = "arn:aws:elasticloadbalancing:***:***:listener/app/***/***/***"

  target_group_arn_web = "arn:aws:elasticloadbalancing:***:***:targetgroup/web/***"
  target_group_arn_blog = "arn:aws:elasticloadbalancing:***:***:targetgroup/blog/***"

  target_group_arn_api1 = "arn:aws:elasticloadbalancing:***:***:targetgroup/api1/***"
  target_group_arn_api2 = "arn:aws:elasticloadbalancing:***:***:targetgroup/api2/***"
}

resource "aws_lb_listener_rule" "web_catch_all" {
  listener_arn = local.load_balancer_listener_arn
  priority = 40901

  action {
    type             = "forward"
    target_group_arn = local.target_group_arn_web
  }

  condition {
    host_header {
      values = ["test.com"]
    }
  }

  condition {
    path_pattern {
      values = ["*"]
    }
  }
}

resource "aws_lb_listener_rule" "api_catch_all" {
  listener_arn = local.load_balancer_listener_arn
  priority = 40902

  action {
    type             = "forward"
    target_group_arn = local.target_group_arn_api1
  }

  condition {
    host_header {
      values = ["api.test.com"]
    }
  }

  condition {
    path_pattern {
      values = ["*"]
    }
  }
}

resource "aws_lb_listener_rule" "web_blog" {
  listener_arn = local.load_balancer_listener_arn

  # would result in a priority of 101
  priority_min = 100
  priority_max = 40900

  action {
    type             = "forward"
    target_group_arn = local.target_group_arn_blog
  }

  condition {
    host_header {
      values = ["test.com"]
    }
  }

  condition {
    path_pattern {
      values = ["/blog*"]
    }
  }
}

resource "aws_lb_listener_rule" "api_2" {
  listener_arn = local.load_balancer_listener_arn

  # would result in a priority of 102
  priority_min = 100
  priority_max = 40900 

  action {
    type             = "forward"
    target_group_arn = local.target_group_arn_api2
  }

  condition {
    host_header {
      values = ["api.test.com"]
    }
  }

  condition {
    path_pattern {
      values = ["/two*"]
    }
  }
}
```

**Comment by garyrutland** [comment]
If deemed useful, this is something I'd be happy to contribute.

**Comment by kennethtxytqw** [comment]
This is needed. It gets really tiresome to handle rule groups (where within a group you dont care about the ordering).

**Comment by ivanrodriguezfernandez** [comment]
This is necessary. Handling rule groups becomes very tedious when the order within the group doesn't matter.

**Comment by xsnrg** [comment]
Yes please! Having this would also solve the problem of reserving high priority rule blocks, say 1-10 for usage such as downtime redirects.

**Comment by jordanst3wart** [comment]
I was just having a read of this issue, and didn't quiet understand it. Could I please get an example, or story of where it is useful?

**Comment by garyrutland** [comment]
I must admit, I didn't realise this issue was still open.
Let me try and remember what my thinking was...

The primary goal here is that, generally speaking, there isn't a need to explicitly set the priority and happy to accept the next available priority but in some situations (thinking maybe if you need to temporarily add in a static response for a maintenance page maybe) you can't do so as all the priorities are starting from 1.
By being able to set a min and possibly a max priority it can give a safe space if you ever need to inject an explicit priority to override something further down.

Hopefully that explains what my use case was back then, or it's just a rehash of my original message.
